src/Core/DaemonClient.php

```diff
@@ -0,0 +1,380 @@
+        self::encodeUnsigned($msg, count($ms));
+        foreach ($ms as $m) {
+            $measure = $m->getMeasure();
+            $msg .= self::encodeString($measure->getName());
+            if ($measure instanceof IntMeasure) {
+                $msg .= self::MS_TYPE_INT;
+                self::encodeUnsigned($msg, $m->getValue());
+            } else if ($measure instanceof FloatMeasure){
+                $msg .= self::MS_TYPE_FLOAT;
+                $msg .= pack('E', $m->getValue());
+            } else {
+                $msg .= self::MS_TYPE_UNKNOWN;
+            }
+        }
+        $tags = $tagContext->tags();
+        self::encodeUnsigned($msg, count($tags));
+        foreach ($tags as $tag) {
+            $msg .= self::encodeString($tag->getKey()->getName());
+            $msg .= self::encodeString($tag->getValue()->getValue());
+        }
+        self::encodeUnsigned($msg, count($attachments));
+        foreach ($attachments as $key => $value) {
+            $msg .= self::encodeString($key);
+            $msg .= self::encodeString($value);
+        }
+        return self::$instance->send(self::MSG_STATS_RECORD, $msg);
+    }
+
+    public function export(array $spans)
+    {
+        $spanData = json_encode(array_map(function (SpanData $span) {
+            return [
+                'traceId' => $span->traceId(),
+                'spanId' => $span->spanId(),
+                'parentSpanId' => $span->parentSpanId(),
+                'name' => $span->name(),
+                'kind' => $span->kind(),
+                'stackTrace' => $span->stackTrace(),
+                'startTime' => $span->startTime(),
+                'endTime' => $span->endTime(),
+                'status' => $span->status(),
+                'attributes' => $span->attributes(),
+                'timeEvents' => $span->timeEvents(),
+                'links' => $span->links(),
+                'sameProcessAsParentSpan' => $span->sameProcessAsParentSpan(),
+            ];
+        }, $spans));
+        $len = '';
+
+        return self::$instance->send(self::MSG_TRACE_EXPORT, $len . $spanData);
+    }
+
+    /**
+     * Send message to daemon
+     *
+     * Message layout:
+     *   MSG_PREFIX  : 32 bit (0 value to aide in recovery from message truncation)
+     *   MESSAGE_TYPE: byte
+     *   SEQUENCE_NR : varint
+     *   PROCESS_ID  : varint
+     *   THREAD_ID   : varint (0 in most PHP deployments - (ZTS disabled))
+     *   START_TIME  : float (measured in microseconds, 32 or 64 bit depending on env.)
+     *   MSG_LEN     : varint (length of message payload)
+     *   MSG         : encoded message of size MSG_LEN
+     *
+     * @param string $type The message type (1 byte).
+     * @param string $msg The message payload.
+     * @return bool Returns true on successful operation.
+     */
+    private final function send(string $type, string $msg = ''): bool
+    {
+        $start = microtime(true);
+        $maxEnd = $start + $this->maxSendTime;
```

Comment:
If we exceed maxSendTime X times during a single request, maybe we should have a per-request circuit-breaker that skips send calls during the rest of the request?

I'm thinking of a case where the agent is hanging and we end up waiting maxSendTime for each measurement.

Reply:
Good idea. I'll add it. Also there will be an extension version of the daemon client which will allow stats recording to be handled outside the php request path.

Reply:
@greggdonovan initial version of php extension powered daemonclient transport can be found in the php-module branch: https://github.com/census-instrumentation/opencensus-php/tree/php-module.